internal/decision/engine.go

```diff
@@ -673,12 +673,18 @@ func (e *Engine) ReceiveFrom(from peer.ID, blks []blocks.Block) {
 	// Check each peer to see if it wants one of the blocks we received
 	var work bool
 	missingWants := make(map[peer.ID][]cid.Cid)
-	e.lock.RLock()
 	for _, b := range blks {
 		k := b.Cid()
 
-		for _, p := range e.peerLedger.Peers(k) {
+		e.lock.RLock()
+		peers := e.peerLedger.Peers(k)
+		e.lock.RUnlock()
+
+		for _, p := range peers {
+			e.lock.RLock()
 			ledger, ok := e.ledgerMap[p]
+			e.lock.RUnlock()
```

Comment:
Does it matter if the operations in the unlocked portion end up happening on an invalid ledger (i.e. one that is either no longer in the map)?